ledger/ledgercore/statedelta.go

```diff
@@ -191,29 +195,87 @@ type AccountDeltas struct {
 	assetResourcesCache map[AccountAsset]int
 }
 
-// MakeStateDelta creates a new instance of StateDelta.
+// MakeStateDelta creates a new instance of StateDelta
 // hint is amount of transactions for evaluation, 2 * hint is for sender and receiver balance records.
 // This does not play well for AssetConfig and ApplicationCall transactions on scale
-func MakeStateDelta(hdr *bookkeeping.BlockHeader, prevTimestamp int64, hint int, stateProofNext basics.Round) StateDelta {
-	return StateDelta{
-		Accts: MakeAccountDeltas(hint),
-		Txids: make(map[transactions.Txid]IncludedTransactions, hint),
-		// asset or application creation are considered as rare events so do not pre-allocate space for them
-		Hdr:                      hdr,
-		StateProofNext:           stateProofNext,
-		PrevTimestamp:            prevTimestamp,
-		initialTransactionsCount: hint,
+func MakeStateDelta(hdr *bookkeeping.BlockHeader, prevTimestamp int64, hint int, stateProofNext basics.Round) (sd StateDelta) {
+	sd = StateDelta{}
+	sd.PopulateStateDelta(hdr, prevTimestamp, hint, stateProofNext)
+	return
+}
+
+// PopulateStateDelta populates an existing StateDelta struct.
+// Used as a helper for MakeStateDelta as well as for re-using already allocated structs from sync.Pool
+func (sd *StateDelta) PopulateStateDelta(hdr *bookkeeping.BlockHeader, prevTimestamp int64, hint int, stateProofNext basics.Round) {
+	if sd.Txids == nil {
+		sd.Txids = make(map[transactions.Txid]IncludedTransactions, hint)
+	}
+	if sd.Accts.isEmpty() {
+		sd.Accts = MakeAccountDeltas(hint)
 	}
+	sd.Hdr = hdr
+	sd.StateProofNext = stateProofNext
+	sd.PrevTimestamp = prevTimestamp
+	sd.initialTransactionsCount = hint
```

Comment:
Good catch on it not making sense outside of the checks. Making a change for that.

I'd be happy to remove the `OptimizedAllocatedMemory` as well but do you have concerns since the sync.Pool implementation will naturally make the hint grow to the largest hint encountered during the lifetime of algod process hence making it more valuable than before even?